aten/src/ATen/tensor_options_utils.py

```diff
@@ -0,0 +1,181 @@
+# These are helper methods around TensorOptions object in codegen
+# While processing native_funtionc.yaml we heavily rely on the parameters types and orders
+# TensorOptions object consists of ScalarType, Layout, Device and Bool (pin_memory) types
+
+# List of tensor options
+tensor_options_args = ['dtype', 'layout', 'device', 'pin_memory']
+
+tensor_options_optional_types_and_var = ['c10::optional<ScalarType> dtype',
+                                         'c10::optional<Layout> layout',
+                                         'c10::optional<Device> device',
+                                         'c10::optional<bool> pin_memory']
+
+tensor_options_types_and_var = ['ScalarType dtype', 'Layout layout', 'Device device', 'bool pin_memory']
+
+# Checks if passed list of arguments contains TensorOptions in it.
+# Returns 'True' in cases when:
+#   1. There is 'TensorOptions' type argument
+#   2. There are all required types of arguments
+#   3. There are all required types of arguments as optional
+def check_if_factory_method(args):
```

Comment:
The existence of this function is supremely confusing to me. My understanding was that one of the original motivations for the TensorOptions removal was to eliminate all of this code that scanned the argument list and reconstructed the TensorOptions argument from it. Therefore, after the refactor, I expect there to not be any code that scans through the argument list looking for "TensorOptions" types, and then reassembling together into a TensorOptions. But that seems to be precisely what this function does.

Let me guess why this function exists (I haven't reviewed the call sites of these functions yet.) Although we have eliminated the internal use of TensorOptions, we still have the problem that we have to generate the user visible C++ bindings that accept TensorOptions.  At this point, we have a Choice(TM)

1. We can type out the C++ bindings by hand. You did that for the registrations and that was kind of not great, so let's say that probably typing out the factory function bindings by hand is also going to be kind of crappy. Maybe you could make it simpler using templates.
2. You can try to reverse engineer the C++ bindings out from the existing `native_functions.yaml`. That's what it looks like this current PR is doing.
3. You can write a small, self contained codegen, with a new data source that provides, in convenient format, the necessary information to do the translation, and then generate C++ bindings. I would seriously consider this option.

I may change my mind as I review more of this patch, but this is my PREDICTION(TM).